File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/overview/InformationActivity.java
package de.tum.in.tumcampusapp.component.ui.overview;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.format.DateUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.other.generic.activity.BaseActivity;
import de.tum.in.tumcampusapp.component.other.reporting.stats.ImplicitCounter;
import de.tum.in.tumcampusapp.utils.Const;
import de.tum.in.tumcampusapp.utils.Utils;

/**
 * Provides information about this app and all contributors
 */
public class InformationActivity extends BaseActivity {

    private final TableRow.LayoutParams rowParams = new TableRow.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
    private int debugOptionsCount;

    public InformationActivity() {
        super(R.layout.activity_information);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ImplicitCounter.count(this);
        this.displayVersionName();
    }

    /**
     * Display version name
     */
    private void displayVersionName() {

        //Get the information
        String versionName = "";
        PackageInfo packageInfo;
        try {
            packageInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            versionName = getResources().getString(R.string.version) + ": " + packageInfo.versionName;
        } catch (NameNotFoundException e) {
            Utils.log(e);
        }

        //Set it up on the ui
        TextView tv = findViewById(R.id.txt_version);
        tv.setText(versionName);

        //Setup showing of debug information
        tv.setOnClickListener(v -> {
            //Lock it after five clicks
            if (debugOptionsCount > 5) {
                return;
            }

            //Increase
            debugOptionsCount++;

            //Show at five clicks
            if (debugOptionsCount == 5) {
                InformationActivity.this.displayDebugInfo();
            }
        });
    }

    private void displayDebugInfo() {

        TableLayout table = findViewById(R.id.debugInfos);
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this);

        this.addDebugRow(table, "LRZ ID", sp.getString(Const.LRZ_ID, ""));
        this.addDebugRow(table, "TUM Access token", sp.getString(Const.ACCESS_TOKEN, ""));
        this.addDebugRow(table, "Bugreports", sp.getBoolean(Const.BUG_REPORTS, false) + " ");
        this.addDebugRow(table, "REG ID", Utils.getInternalSettingString(this, Const.GCM_REG_ID, ""));
        this.addDebugRow(table, "REG Transmission", DateUtils.getRelativeDateTimeString(this, Utils.getInternalSettingLong(this, Const.GCM_REG_ID_LAST_TRANSMISSION, 0),
                                                                                        DateUtils.MINUTE_IN_MILLIS, DateUtils.DAY_IN_MILLIS * 2, 0)
                                                             .toString());
        try {
            PackageInfo packageInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            this.addDebugRow(table, "VersionCode", String.valueOf(packageInfo.versionCode));
        } catch (NameNotFoundException e) {
            Utils.log(e);
        }

        table.setVisibility(View.VISIBLE);

    }

    private void addDebugRow(TableLayout t, final String label, final String value) {
        //Create new row
        TableRow tableRow = new TableRow(this);
        tableRow.setLayoutParams(rowParams);

        //Add our text fields
        TextView l = new TextView(this);
        l.setText(label);
        l.setLayoutParams(rowParams);
        tableRow.addView(l);

        TextView v = new TextView(this);
        v.setText(value);
        v.setLayoutParams(rowParams);
        v.setClickable(true);

        //Copy to clipboard
        v.setOnClickListener(v1 -> {
            ClipboardManager clipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText(label, value);
            clipboard.setPrimaryClip(clip);
        });
        tableRow.addView(v);

        //Add it to the table
        t.addView(tableRow);
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/overview/InformationActivity.kt
package de.tum.`in`.tumcampusapp.component.ui.overview

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Intent
import android.content.Intent.ACTION_VIEW
import android.content.pm.PackageManager.NameNotFoundException
import android.net.Uri
import android.os.Bundle
import androidx.preference.PreferenceManager
import android.text.format.DateUtils
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import androidx.core.content.pm.PackageInfoCompat
import de.psdev.licensesdialog.LicensesDialog
import de.tum.`in`.tumcampusapp.BuildConfig
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.other.generic.activity.BaseActivity
import de.tum.`in`.tumcampusapp.databinding.ActivityInformationBinding
import de.tum.`in`.tumcampusapp.utils.Const
import de.tum.`in`.tumcampusapp.utils.Utils

/**
 * Provides information about this app and all contributors
 */
class InformationActivity : BaseActivity(R.layout.activity_information) {

    private lateinit var binding: ActivityInformationBinding
    private val rowParams = TableRow.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityInformationBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.buttonFacebook.setOnClickListener {
            openFacebook()
        }
        binding.buttonGithub.setOnClickListener {
            startActivity(Intent(ACTION_VIEW, Uri.parse(getString(R.string.github_link))))
        }
        binding.buttonPrivacy.setOnClickListener {
            startActivity(Intent(ACTION_VIEW, Uri.parse(getString(R.string.url_privacy_policy))))
        }
        binding.buttonLicenses.setOnClickListener {
            LicensesDialog.Builder(this)
                    .setNotices(R.raw.notices)
                    .setShowFullLicenseText(false)
                    .setIncludeOwnLicense(true)
                    .build()
                    .show()
        }


        displayDebugInfo()
    }

    /**
     * Open the Facebook app or view page in a browser if Facebook is not installed.
     */
    private fun openFacebook() {
        try {
            packageManager.getPackageInfo("com.facebook.katana", 0)
            val intent = Intent(ACTION_VIEW, Uri.parse(getString(R.string.facebook_link_app)))
            startActivity(intent)
        } catch (e: Exception) {
            // Don't make any assumptions about another app, just start the browser instead
            val default = Intent(ACTION_VIEW, Uri.parse(getString(R.string.facebook_link)))
            startActivity(default)
        }
    }

    private fun displayDebugInfo() {
        // Setup showing of debug information
        val sp = PreferenceManager.getDefaultSharedPreferences(this)

        with(binding) {
            try {
                val packageInfo = packageManager.getPackageInfo(packageName, 0)
                addDebugRow(debugInfos, "App version", packageInfo.versionName)
            } catch (ignore: NameNotFoundException) {
            }

            addDebugRow(debugInfos, "TUM ID", sp.getString(Const.LRZ_ID, ""))
            val token = sp.getString(Const.ACCESS_TOKEN, "")
            if (token == "") {
                addDebugRow(debugInfos, "TUM access token", "")
            } else {
                addDebugRow(debugInfos, "TUM access token", token?.substring(0, 5) + "...")
            }
            addDebugRow(debugInfos, "Bug reports", sp.getBoolean(Const.BUG_REPORTS, false).toString() + " ")

            addDebugRow(debugInfos, "REG ID", Utils.getSetting(applicationContext, Const.FCM_REG_ID, ""))
            addDebugRow(debugInfos, "REG transmission", DateUtils.getRelativeDateTimeString(applicationContext,
                    Utils.getSettingLong(applicationContext, Const.FCM_REG_ID_LAST_TRANSMISSION, 0),
                    DateUtils.MINUTE_IN_MILLIS, DateUtils.DAY_IN_MILLIS * 2, 0).toString())
            try {
                val packageInfo = packageManager.getPackageInfo(packageName, 0)
                addDebugRow(debugInfos, "Version code", PackageInfoCompat.getLongVersionCode(packageInfo).toString())
            } catch (ignore: NameNotFoundException) {
            }

            addDebugRow(debugInfos, "Build configuration", BuildConfig.DEBUG.toString())
            debugInfos.visibility = View.VISIBLE
        }

    }

    private fun addDebugRow(tableLayout: TableLayout, label: String, value: String?) {
        val tableRow = TableRow(this)
        tableRow.layoutParams = rowParams

        val labelTextView = TextView(this).apply {
            text = label
            layoutParams = rowParams

            val padding = resources.getDimensionPixelSize(R.dimen.material_small_padding)
            setPadding(0, 0, padding, 0)
        }
        tableRow.addView(labelTextView)

        val valueTextView = TextView(this).apply {
            text = value
            layoutParams = rowParams
            isClickable = true

            setOnLongClickListener {
                val clipboard = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
                val clip = ClipData.newPlainText(label, value)
                clipboard.setPrimaryClip(clip)
                true
            }
        }
        tableRow.addView(valueTextView)

        tableLayout.addView(tableRow)
    }
}
